Hipcms 682

Musaevt opened the pull request:
## Added API Calls:
* GET /api/Exhibits/rating/{id}
* POST GET /api/Exhibits/rating/{id}

#### Added Raiting System that can be applied to any ResourceType

Musaevt committed:
Added Rating System
--Rating system can be aplied to any ResourceType
--Added rating to Exhibits

Musaevt committed:
Merge branch 'develop' into HIPCMS-682

Musaevt committed:
Merge branch 'develop' into HIPCMS-682

# Conflicts:
#	HiP-DataStore/Controllers/ExhibitsController.cs

Musaevt committed:
Added minor cleanup

Musaevt committed:
R# Error fixing

Musaevt committed:
R# ResourceType  redundance fixing

SvenEV commented on HiP-DataStore.Model/Entity/Rating.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public class Rating
+    {
+        public int Id;
```
You are using public fields here instead of properties. Please use properties (e.g. `public int Id { get; set; }`).

SvenEV commented on HiP-DataStore.Model/ResourceType.cs:
```diff
@@ -3,14 +3,15 @@
 namespace PaderbornUniversity.SILab.Hip.DataStore.Model
 {
-    public struct ResourceType : IEquatable<ResourceType>
+    public class ResourceType : IEquatable<ResourceType>
```
Why did you need to change the `struct` to a `class`?

SvenEV commented on HiP-DataStore/Controllers/ExhibitsController.cs:
```diff
@@ -193,6 +195,73 @@ public IActionResult GetById(int id, DateTimeOffset? timestamp = null)
+        [ProducesResponseType(typeof(int),201)]
+        [ProducesResponseType(400)]
+        [ProducesResponseType(404)]
+        public async Task<IActionResult> PostRating(int id,RatingArgs args)
```
Should be called `PostRatingAsync`

SvenEV commented on HiP-DataStore/Controllers/ExhibitsController.cs:
```diff
@@ -193,6 +195,73 @@ public IActionResult GetById(int id, DateTimeOffset? timestamp = null)
             return NoContent();
         }
 
+        [HttpGet("rating/{id}")]
```
Although it doesn't really matter I have been told before that our HTTP routes should begin with upper case letters, i.e. "Rating/{id}"

SvenEV commented on HiP-DataStore/Controllers/ExhibitsController.cs:
```diff
@@ -193,6 +195,73 @@ public IActionResult GetById(int id, DateTimeOffset? timestamp = null)
+            var oldValue = _db.Database
+                              .GetCollection<Rating>(ResourceType.Rating.Name)
+                              .AsQueryable()
+                              .Where(x => x.UserId == args.UserId.GetValueOrDefault() && x.RatedType.Name == ResourceType.Exhibit.Name && x.EntityId == id)
```
You can directly check `x.RatedType == ResourceType.Exhibit`, leaving out `.Name` on both sides

SvenEV commented on HiP-DataStore/Controllers/ExhibitsController.cs:
```diff
@@ -193,6 +195,73 @@ public IActionResult GetById(int id, DateTimeOffset? timestamp = null)
+            };
+
+            await _eventStore.AppendEventAsync(ev);
+            return Created($"{Request.Scheme}://{Request.Host}/api/Exhibits/{ev.Id}", ev.Id);
```
Incorrect URL returned here

SvenEV commented on HiP-DataStore/Controllers/ExhibitsController.cs:
```diff
@@ -193,6 +195,73 @@ public IActionResult GetById(int id, DateTimeOffset? timestamp = null)
+            if (!ModelState.IsValid)
+                return BadRequest(ModelState);
+
+            var exhibit = _db.Database.GetCollection<Exhibit>(ResourceType.Exhibit.Name)
```
Remember we shouldn't rely on the MongoDB database when handling PUT- or POST-requests as it may not be consistent with the Event Store. Normally, things go like this:
- user posts a rating of 3 (old rating = null) => RatingAdded-event <span>#</span>1 is emitted
- Mongo cache DB receives RatingAdded-event <span>#</span>1
- user posts a rating of 4 (old rating = 3) => RatingAdded-event <span>#</span>2 is emitted
- Mongo cache DB receives RatingAdded-event <span>#</span>2

But the order is not predictable, so something like this may also happen:
- user posts a rating of 3 (old rating = null) => RatingAdded-event <span>#</span>1 is emitted
- user posts a rating of 4 (**old rating = null**) => RatingAdded-event <span>#</span>2 is emitted
- Mongo cache DB receives RatingAdded-event <span>#</span>1
- Mongo cache DB receives RatingAdded-event <span>#</span>2

...and this may mess up your rating calculations.

SvenEV reviewed (changes requested):
Nice work overall (especially since it's easily applicable to all resource types), but some things need to be addressed before merging into develop.

Musaevt committed:
Added changes:
--changed all url form 'rating' to 'Rating'
--public fields instead of properties.
--changed URL for exhibit/rating

Musaevt committed:
--Added savings of all user`s ratings in memory.
--Changed rating type to byte

Musaevt committed:
Merge branch 'develop' into HIPCMS-682

# Conflicts:
#	HiP-DataStore/Controllers/ExhibitsController.cs

Musaevt committed:
R# Error fixing

SvenEV commented on HiP-DataStore.Model/Rest/RatingArgs.cs:
```diff
@@ -0,0 +1,16 @@
+    public class RatingArgs
+    {
+        [Required]
+        [Range(1, 5, ErrorMessage = "Has to be in a range from 1 to 5")]
```
Perhaps use `RatingEntityInfo.MinRateValue` and `RatingEntityInfo.MaxRateValue` here?

SvenEV commented on HiP-DataStore.Model/Entity/Rating.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace PaderbornUniversity.SILab.Hip.DataStore.Model.Entity
+{
+    public class Rating
```
Note that we are no longer querying the Mongo database for rating-related things (since it's all in RAM)... so we don't even need to store ratings in the Mongo database, right?

SvenEV replied:
On the other hand... a cleaner Command & Query separation could be achieved if we query the database in `ExhibitsController.GetRating(int)` (instead of using `RatingIndex`). What do you think?

Musaevt committed:
No Storing Ratings in storage
Min-Max Rating now can be defined in one place
